core/controllers/classifier.py

```diff
@@ -0,0 +1,148 @@
+    job_id = message.get('job_id')
+    classifier_data = message.get('classifier_data')
+
+    if not isinstance(job_id, basestring):
+        return False
+    if not isinstance(classifier_data, dict):
+        return False
+    return True
+
+
+def verify_signature(message, vm_id, received_signature):
+    """Function that checks if the signature received from the VM is valid.
+
+    Args:
+        message: dict. The message payload data.
+        vm_id: str. The ID of the VM instance.
+        received_signature: str. The signature received from the VM.
+
+    Returns:
+        bool. Returns True if the incoming request is valid, False otherwise.
+
+    Raises:
+        UnauthorizedUserException. Exception is raised when the incoming
+            request is invalid.
+    """
+    secret = None
+    for val in config_domain.VMID_SHARED_SECRET_KEY_MAPPING.value:
+        if val['vm_id'] == vm_id:
+            secret = str(val['shared_secret_key'])
+            break
+    if secret is None:
+        return False
+
+    generated_signature = generate_signature(secret, message)
+    if generated_signature != received_signature:
+        return False
+    return True
+
+
+class TrainedClassifierHandler(base.BaseHandler):
+    """This handler stores the result of the training job in datastore and
+    updates the status of the job.
+    """
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    def post(self):
+        """Handles POST requests."""
+        signature = self.payload.get('signature')
+        message = self.payload.get('message')
+        vm_id = self.payload.get('vm_id')
+        if vm_id == feconf.DEFAULT_VM_ID and not feconf.DEV_MODE:
+            raise self.UnauthorizedUserException
+
+        if not validate_message(message):
+            raise self.InvalidInputException
+        if not verify_signature(message, vm_id, signature):
+            raise self.UnauthorizedUserException
+
+        job_id = message.get('job_id')
+        classifier_data = message.get('classifier_data')
+        classifier_training_job = (
+            classifier_services.get_classifier_training_job_by_id(job_id))
+        state_name = classifier_training_job.state_name
+        exp_id = classifier_training_job.exp_id
+        exp_version = classifier_training_job.exp_version
+        algorithm_id = classifier_training_job.algorithm_id
+        interaction_id = classifier_training_job.interaction_id
+
+        data_schema_version = None
+        if feconf.INTERACTION_CLASSIFIER_MAPPING[interaction_id][
+                'algorithm_id'] == algorithm_id:
+            data_schema_version = feconf.INTERACTION_CLASSIFIER_MAPPING[
+                interaction_id]['current_data_schema_version']
+        if data_schema_version is None:
+            raise self.InternalErrorException(
+                'The algorithm_id of the job does not exist in the Interaction'
+                'Classifier Mapping.')
+
+        classifier = classifier_domain.ClassifierData(
```

Comment:
I don't think you should construct this here -- too much logic. Try `classifier_services.create_classifier(job_id, classifier_data)` instead.

(Sorry if I miswrote in the previous review.)